File: q0452/q0452.py
class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        points.sort()
        total = 1
        prevEnd = points[0][1]
        for i in range(1, len(points)):
            currStart, currEnd = points[i]
            
            # If this current start is before previous end, there is no need to shoot another arrow
            # Update the previous end in case the current end is smaller
            if currStart <= prevEnd: 
                prevEnd = min(prevEnd, currEnd)

            # If this current start is after previous end, then we need to shoot one arrow at the end of previous end
            # Update the previous end to be the value of current end
            else:
                prevEnd = currEnd
                total += 1
        return total

File: q0452/q0452.java
class Solution {
    public int findMinArrowShots(int[][] points) {
        Arrays.sort(points, (p1, p2) -> Integer.compare(p1[0], p2[0]));
        int total = 1;
        int prevEnd = points[0][1];
        for (int i = 1; i < points.length; i++) {
            int currStart = points[i][0];
            int currEnd = points[i][1];
            if (currStart <= prevEnd)
                prevEnd = Math.min(prevEnd, currEnd);
            else {
                total += 1;
                prevEnd = currEnd;
            }
        }
        return total;
    }
}